File: ios/Calc/Models/EmulatorPreferences.swift
//
//  EmulatorPreferences.swift
//  Calc
//
//  Preferences helper for emulator settings.
//  Stores user preferences in UserDefaults.
//

import Foundation

/// Manages persistent emulator preferences.
class EmulatorPreferences {
    // MARK: - Keys

    private static let backendKey = "preferredBackend"
    private static let speedMultiplierKey = "speedMultiplier"
    private static let autoSaveKey = "autoSaveEnabled"
    private static let lastRomHashKey = "lastRomHash"
    private static let lastRomNameKey = "lastRomName"
    private static let calculatorScaleKey = "calculatorScale"
    private static let calculatorYOffsetKey = "calculatorYOffset"

    // MARK: - Backend

    /// Get the user's preferred backend name.
    /// - Returns: Backend name or nil to use default
    static func getPreferredBackend() -> String? {
        return UserDefaults.standard.string(forKey: backendKey)
    }

    /// Set the user's preferred backend name.
    /// - Parameter name: Backend name to save
    static func setPreferredBackend(_ name: String) {
        UserDefaults.standard.set(name, forKey: backendKey)
    }

    /// Get the effective backend to use.
    /// Returns the preferred backend if available, otherwise the first available backend.
    /// - Returns: Backend name to use, or nil if no backends available
    static func getEffectiveBackend() -> String? {
        let preferred = getPreferredBackend()
        let available = EmulatorBridge.getAvailableBackends()

        // If preferred backend is available, use it
        if let preferred = preferred, available.contains(preferred) {
            return preferred
        }

        // Otherwise use the first available
        return available.first
    }

    // MARK: - Speed

    /// Speed multiplier (default 1.0)
    static var speedMultiplier: Float {
        get {
            let value = UserDefaults.standard.float(forKey: speedMultiplierKey)
            return value > 0 ? value : 1.0
        }
        set {
            UserDefaults.standard.set(newValue, forKey: speedMultiplierKey)
        }
    }

    // MARK: - Display

    /// Calculator scale (default 1.0)
    static var calculatorScale: Float {
        get {
            let value = UserDefaults.standard.float(forKey: calculatorScaleKey)
            return value > 0 ? value : 1.0
        }
        set {
            UserDefaults.standard.set(newValue, forKey: calculatorScaleKey)
        }
    }

    /// Calculator Y offset (default 0)
    static var calculatorYOffset: Float {
        get { UserDefaults.standard.float(forKey: calculatorYOffsetKey) }
        set { UserDefaults.standard.set(newValue, forKey: calculatorYOffsetKey) }
    }

    // MARK: - Auto-save

    /// Whether auto-save is enabled (default true)
    static var autoSaveEnabled: Bool {
        get {
            // Default to true if not set
            if UserDefaults.standard.object(forKey: autoSaveKey) == nil {
                return true
            }
            return UserDefaults.standard.bool(forKey: autoSaveKey)
        }
        set {
            UserDefaults.standard.set(newValue, forKey: autoSaveKey)
        }
    }

    // MARK: - Last ROM

    /// Hash of the last loaded ROM
    static var lastRomHash: String? {
        get { UserDefaults.standard.string(forKey: lastRomHashKey) }
        set { UserDefaults.standard.set(newValue, forKey: lastRomHashKey) }
    }

    /// Name of the last loaded ROM
    static var lastRomName: String? {
        get { UserDefaults.standard.string(forKey: lastRomNameKey) }
        set { UserDefaults.standard.set(newValue, forKey: lastRomNameKey) }
    }

    // MARK: - Clear

    /// Clear all saved ROM/state info (for fresh start)
    static func clearLastRomInfo() {
        lastRomHash = nil
        lastRomName = nil
    }
}

File: android/app/src/main/java/com/calc/emulator/EmulatorPreferences.kt
package com.calc.emulator

import android.content.Context
import android.content.SharedPreferences

/**
 * Preferences helper for emulator settings.
 */
object EmulatorPreferences {
    private const val PREFS_NAME = "emulator_prefs"
    private const val KEY_BACKEND = "backend"
    private const val KEY_LAST_ROM_HASH = "last_rom_hash"
    private const val KEY_CALCULATOR_SCALE = "calculator_scale"
    private const val KEY_CALCULATOR_Y_OFFSET = "calculator_y_offset"

    private fun getPrefs(context: Context): SharedPreferences {
        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    /**
     * Get the preferred backend name.
     * @return Backend name or null to use default
     */
    fun getPreferredBackend(context: Context): String? {
        return getPrefs(context).getString(KEY_BACKEND, null)
    }

    /**
     * Set the preferred backend name.
     * @param backendName Backend name to save
     */
    fun setPreferredBackend(context: Context, backendName: String) {
        getPrefs(context).edit().putString(KEY_BACKEND, backendName).apply()
    }

    /**
     * Get the effective backend to use.
     * Returns the preferred backend if available, otherwise the first available backend.
     */
    fun getEffectiveBackend(context: Context): String? {
        val preferred = getPreferredBackend(context)
        val available = EmulatorBridge.getAvailableBackends()

        // If preferred backend is available, use it
        if (preferred != null && available.contains(preferred)) {
            return preferred
        }

        // Otherwise use the first available
        return available.firstOrNull()
    }

    /**
     * Get the last used ROM hash.
     * @return ROM hash or null if no ROM was previously loaded
     */
    fun getLastRomHash(context: Context): String? {
        return getPrefs(context).getString(KEY_LAST_ROM_HASH, null)
    }

    /**
     * Set the last used ROM hash.
     * @param hash ROM hash to save
     */
    fun setLastRomHash(context: Context, hash: String) {
        getPrefs(context).edit().putString(KEY_LAST_ROM_HASH, hash).apply()
    }

    fun getCalculatorScale(context: Context): Float {
        val value = getPrefs(context).getFloat(KEY_CALCULATOR_SCALE, 1f)
        return if (value > 0) value else 1f
    }

    fun setCalculatorScale(context: Context, scale: Float) {
        getPrefs(context).edit().putFloat(KEY_CALCULATOR_SCALE, scale).apply()
    }

    fun getCalculatorYOffset(context: Context): Float {
        return getPrefs(context).getFloat(KEY_CALCULATOR_Y_OFFSET, 0f)
    }

    fun setCalculatorYOffset(context: Context, offset: Float) {
        getPrefs(context).edit().putFloat(KEY_CALCULATOR_Y_OFFSET, offset).apply()
    }

    /**
     * Clear the last ROM hash (e.g., when ROM fails to load).
     */
    fun clearLastRomHash(context: Context) {
        getPrefs(context).edit().remove(KEY_LAST_ROM_HASH).apply()
    }
}
